### PR TITLE
Perform CI builds on macOS 14 for ARM64 and macOS 13 for x86_64.

### PR DESCRIPTION
**Pull request by tmcgilchrist**
This PR makes 3 changes:

1. Adds macsOS 14 ARM64 runner under the name `macos-arm64`, which we didn't previously test on.
1. Updates macOS x86_64 to 13 from 12 under the name `macOS-x86_64`
1. Updates the version of actions/checkout to v4

See GitHub announcement here https://github.blog/changelog/2024-01-30-github-actions-macos-14-sonoma-is-now-available/

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -150,8 +150,10 @@ jobs:
           - name: linux-O0
             os: ubuntu-latest
             config_arg: CFLAGS='-O0'
-          - name: macos
-            os: macos-latest
+          - name: macos-x86_64
+            os: macos-14-large
```
Comment (anmonteiro) [inline .github/workflows/build.yml]:
```suggestion
            os: macos-13
```

Comment (anmonteiro) [reply .github/workflows/build.yml]: from the blog post:

> The macOS 12 runner image will remain latest until migration of the latest YAML workflow label to macOS 14 in Q2 FY24 (April – June 2024).

Comment (anmonteiro) [reply .github/workflows/build.yml]: and 

> While macOS 13 is now generally available under the macos-13 label, this image will not be migrated to latest.

perhaps this should be 13 so we make sure to stay on x86_64?

Comment (tmcgilchrist) [reply .github/workflows/build.yml]: `macos-14-large` seems to be x86_64 in my testing while `macos-14` is ARM64.
See the Checks performed on https://github.com/tarides/runtime_events_tools/pull/42.

Comment (anmonteiro) [reply .github/workflows/build.yml]: oh wow, that's confusing. indeed, from https://github.com/actions/runner-images/tree/main?tab=readme-ov-file#available-images you're right

image

Comment (tmcgilchrist) [reply .github/workflows/build.yml]: The real joy is in debugging the yaml 😬 

I don't personally see much point in keeping macOS 12 or 13 around in the build matrix, but others might disagree.

Comment (tmcgilchrist) [reply .github/workflows/build.yml]: Good news is `macos-13` can stay on as an x86_64 build. 
Bad news is `macos-14-large` seems to cost more than the plan present on `ocaml` account in GitHub. Not sure what the cost / plan is, perhaps @avsm might know.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -150,15 +150,17 @@ jobs:
...
+            os: macos-14
     steps:
       - name: Checkout
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
         with:
           persist-credentials: false
       - name: OS Dependencies
-        if: runner.os == 'MacOS'
+        if: ${{ startsWith(matrix.os, 'macos') }}
```
Comment (dra27) [inline .github/workflows/build.yml]: The string for `runner.os` I think is actually `macOS` (although I presume it doesn't matter), but otherwise what's the reason for this change?

**Review by dra27 — COMMENTED** [review]
Small drive-by

File: .github/workflows/build.yml (hunk 2 above)
Comment (tmcgilchrist) [reply .github/workflows/build.yml]: I thought it was more obvious to base this on the `matrix.os` value rather than `runner.os`, since we set that a few lines above. `startsWith` is [not case sensitive](https://docs.github.com/en/actions/learn-github-actions/expressions#startswith).
However it works without this change so it can be dropped.

Comment (dra27) [reply .github/workflows/build.yml]: I think the fact we need to install dependencies is a property of the OS it's physically running on, not what we declare in the matrix (so I'd prefer to keep it based on `runner.os`)

Comment (tmcgilchrist) [reply .github/workflows/build.yml]: Make sense I've removed that change.

**Review by dra27 — APPROVED** [review]
LGTM, thank you!

**Comment by tmcgilchrist** [comment]
Thanks @dra27
